### PR TITLE
feat: make pnpm wait on lock

### PR DESCRIPTION
**Pull request by andreypopp**
Also use `proper-lockfile` package. A similar thing which is yarn does.

### CODE REVIEW

File: src/api/lock.ts [hunk 1]
```diff
@@ -1,18 +1,54 @@
...
+async function lock(lockFilename: string, notFirstTime: boolean = false): Promise<{}> {
+  const promise = new Promise((resolve, reject) => {
+    lockfile.lock(
+      lockFilename,
+      {realpath: false},
+      async (err: Error & {code: string}) => {
+        if (err) {
+          if (err.code === 'ELOCKED') {
+            if (!notFirstTime) {
+              logger.warn('waiting for another installation to complete...')
```
Comment (andreypopp) [inline src/api/lock.ts]: I'm sure I didn't get it right — user don't see a message...

Comment (zkochan) [reply src/api/lock.ts]: he/she should. Warnings are printed. Here is the related code https://github.com/pnpm/pnpm-reporter-default/blob/master/src/index.ts#L45-L48

File: src/api/lock.ts [hunk 2]
```diff
@@ -1,18 +1,54 @@
...
+import lockfile = require('proper-lockfile')
 
-export default async function<T> (storePath: string, fn: () => Promise<T>): Promise<T> {
-  const lockfile: string = path.resolve(storePath, 'lock')
-  await lock(lockfile)
+async function delay(ms: number) {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+async function lock(lockFilename: string, notFirstTime: boolean = false): Promise<{}> {
```
Comment (zkochan) [inline src/api/lock.ts]: It is better to name booleans without negation. So `firstTime: boolean = true` in this case

File: src/api/lock.ts [hunk 3]
```diff
@@ -1,18 +1,54 @@
...
+  const promise = new Promise((resolve, reject) => {
+    lockfile.lock(
+      lockFilename,
+      {realpath: false},
+      async (err: Error & {code: string}) => {
+        if (err) {
```
Comment (zkochan) [inline src/api/lock.ts]: I prefer when nesting is avoided, but I don't want to start a holy war, so I don't insist.

File: src/api/lock.ts (hunk 2 above)
Comment (zkochan) [reply src/api/lock.ts]: Also seems like Node 4 tests don't pass. Probably because of the default argument value

File: src/api/lock.ts (hunk 1 above)
Comment (andreypopp) [reply src/api/lock.ts]: Yep, seen this but during tests they are not printed for some reason.

Comment (zkochan) [reply src/api/lock.ts]: that is expected. bole is used for logging and it should be configured to print output to the console. I intentionally did it silent to have less noise in tests

**Comment by andreypopp** [comment]
Addressed the feedback

**Pull request merged**